### PR TITLE
SONAR-8454 Background task failure on project kee unicity violation after SQ 5.6 upgrade

### PR DESCRIPTION
**Pull request by simonbrandhof**
https://jira.sonarsource.com/browse/SONAR-8454

### CODE REVIEW

File: sonar-db/src/main/java/org/sonar/db/version/v564/CleanUsurperRootComponents.java [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+    cleanSnapshotWithIncorrectRoot(context);
+  }
+
+  private static void fixSnapshotScopeAndQualifier(Context context) throws SQLException {
+    MassUpdate massUpdate = context.prepareMassUpdate();
+    massUpdate.select("select" +
+      " sn.id,p.scope,p.qualifier" +
+      " from" +
+      " snapshots sn, projects p" +
+      " where" +
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/version/v564/CleanUsurperRootComponents.java]: CRITICAL Define a constant instead of duplicating this literal " where" 3 times. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1192)

File: sonar-db/src/main/java/org/sonar/db/version/v564/CleanUsurperRootComponents.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+          update.setString(1, componentUuid);
+          update.setString(2, componentUuid);
+          return true;
+        case 8:
+        case 9:
+        case 10:
+        case 11:
+          update.setLong(1, componentId);
+          return true;
+        case 12:
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/version/v564/CleanUsurperRootComponents.java]: MINOR This case's code block is the same as the block for the case on line 94. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1871)

File: sonar-db/src/main/java/org/sonar/db/version/v564/CleanUsurperRootComponents.java [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+    massUpdate.update("delete from snapshots where component_uuid=? or root_component_uuid=?");
+    massUpdate.update("delete from issues where component_uuid=? or project_uuid=?");
+    massUpdate.update("delete from file_sources where file_uuid=? or project_uuid=?");
+    massUpdate.update("delete from group_roles where resource_id=?");
+    massUpdate.update("delete from user_roles where resource_id=?");
+    massUpdate.update("delete from properties where resource_id=?");
+    massUpdate.update("delete from widgets where resource_id=?");
+    massUpdate.update("delete from projects where uuid=?");
+    massUpdate.rowPluralName("usurper root components");
+    massUpdate.execute((row, update, updateIndex) -> {
```
Comment (SonarTech) [inline sonar-db/src/main/java/org/sonar/db/version/v564/CleanUsurperRootComponents.java]: MAJOR Reduce this lambda expression number of lines from 30 to at most 20. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1188)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* CRITICAL 2 critical
* MAJOR 2 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [ws](https://github.com/SonarSource/sonarqube/blob/9325f07cb9bef05a9bde61e93a2bba9ff1407c2d/server/sonar-server/src/main/java/org/sonar/server/updatecenter/ws): Add a 'package-info.java' file to document the 'ws' package [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1228)

**Comment by SonarTech** [comment]
SonarQube analysis reported 5 issues
* CRITICAL 2 critical
* MAJOR 2 major
* MINOR 1 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [ws](https://github.com/SonarSource/sonarqube/blob/38b452d9436f90d5e1eec694ab6260ef6999783e/server/sonar-server/src/main/java/org/sonar/server/updatecenter/ws): Add a 'package-info.java' file to document the 'ws' package [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1228)

**Comment by sns-seb** [comment]
commit title `SONAR-8454 Background task failure on project kee unicity violation after SQ 5.6 upgrade" too long
describe content of the commit instead

File: sonar-db/src/main/java/org/sonar/db/version/MigrationStepModule.java [hunk 4]
```diff
@@ -171,6 +172,7 @@ protected void configureModule() {
       UpdateUsersExternalIdentityWhenEmpty.class,
 
       // 5.6.4
+      FixProjectUuidOfDeveloperProjects.class,
```
Comment (sns-seb) [inline sonar-db/src/main/java/org/sonar/db/version/MigrationStepModule.java]: merging into 6.2 or master, you should make sure these two classes do not appear twice in the module

Comment (sns-seb) [inline sonar-db/src/main/java/org/sonar/db/version/MigrationStepModule.java]: best is to use strategy "ours" to merge this commit, as this is a backport

File: travis.sh [hunk 5]
```diff
@@ -4,7 +4,7 @@ set -euo pipefail
 
 function configureTravis {
   mkdir ~/.local
-  curl -sSL https://github.com/SonarSource/travis-utils/tarball/v31 | tar zx --strip-components 1 -C ~/.local
+  curl -sSL https://github.com/SonarSource/travis-utils/tarball/v33 | tar zx --strip-components 1 -C ~/.local
```
Comment (sns-seb) [inline travis.sh]: any real benefit to upgrade to 33 in branch-5.6?

Comment (simonbrandhof) [inline travis.sh]: this was a (bad) attempt to enable QA. "DO NOT MERGE" message is missing here.

**Comment by sns-seb** [comment]
I withdraw my approval. The backported migration is failing on 5.6:
{code}
2016.12.01 14:36:36 ERROR web[o.s.s.d.m.DatabaseMigrator] Fail to execute database migration: org.sonar.db.version.v564.CleanUsurperRootComponents
java.sql.SQLSyntaxErrorException: ORA-00904: "SN"."COMPONENT_UUID": invalid identifier
{code}

This is because tables  project_measures et snapshots do not have the UUID column in 5.6

the reproducer is simple: start SQ on a blank Oracle DB

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues
* CRITICAL 1 critical
* MAJOR 3 major
* MINOR 2 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [ws](https://github.com/SonarSource/sonarqube/blob/63c9c81d022b6eb7d450b1d1a5a565caf83a74c7/server/sonar-server/src/main/java/org/sonar/server/updatecenter/ws): Add a 'package-info.java' file to document the 'ws' package [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1228)

File: travis.sh [hunk 6]
```diff
@@ -55,16 +55,19 @@ CI)
     # No need for Maven phase "install" as the generated JAR file does not need to be installed
     # in Maven local repository. Phase "verify" is enough.
 
+    set_maven_build_version $TRAVIS_BUILD_NUMBER
```
Comment (sns-seb) [inline travis.sh]: shouldn't these changes to `travis.sh` be isolated into a dedicated commit?

Comment (simonbrandhof) [inline travis.sh]: Good catch. Obviously you're right !

Comment (simonbrandhof) [inline travis.sh]: Change extracted.

**Comment by SonarTech** [comment]
SonarQube analysis reported 6 issues
* CRITICAL 1 critical
* MAJOR 3 major
* MINOR 2 minor

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MAJOR [ws](https://github.com/SonarSource/sonarqube/blob/c2acafee36d2dbf51e5ab244a7d6f48a01c62243/server/sonar-server/src/main/java/org/sonar/server/updatecenter/ws): Add a 'package-info.java' file to document the 'ws' package [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1228)
